File: server/service/agent/fm.go
package agent

// fm.go — Agent File Manager 服务层。
// 通过 WebSocket 控制通道向 Agent 发送 FM 请求，获取文件列表、下载/上传/删除文件、创建目录。
//
// 协议（与 agent 侧 ws_client/handler.rs 中 FM 处理对称）：
//   Controller → Agent (text JSON): fm_list    { id, payload: { path } }
//   Controller → Agent (text JSON): fm_download { id, payload: { path } }
//   Controller → Agent (text JSON): fm_upload   { id, payload: { path, data(base64), size } }
//   Controller → Agent (text JSON): fm_delete   { id, payload: { path } }
//   Controller → Agent (text JSON): fm_mkdir    { id, payload: { path } }
//   Agent → Controller (text JSON): fm_list_resp / fm_download_resp / fm_upload_resp /
//                                    fm_delete_resp / fm_mkdir_resp / fm_error

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"

	"oneclickvirt/global"

	"go.uber.org/zap"
)

const (
	fmDefaultTimeout = 60 * time.Second
	fmUploadTimeout  = 120 * time.Second
	// 单次上传文件大小上限（50 MB），超过此大小拒绝请求以保护 WebSocket 通道。
	fmMaxUploadBytes = 50 * 1024 * 1024
)

// sendFMRequest 向 Agent 发送 FM 请求帧，并等待响应。
// 返回 (响应消息类型, 原始 payload, error)。
func (a *AgentConn) sendFMRequest(msgType string, payload interface{}, timeout time.Duration) (string, json.RawMessage, error) {
	reqID := randomID()

	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return "", nil, fmt.Errorf("序列化 FM 请求失败: %w", err)
	}

	msg := wsMessage{
		Type:    msgType,
		ID:      reqID,
		Payload: payloadBytes,
	}
	raw, err := json.Marshal(msg)
	if err != nil {
		return "", nil, fmt.Errorf("序列化 FM 消息失败: %w", err)
	}

	respCh := make(chan fmRawResp, 1)
	a.mu.Lock()
	// 检查连接是否已断开
	select {
	case <-a.doneCh:
		a.mu.Unlock()
		return "", nil, fmt.Errorf("Agent 已断开连接")
	default:
	}
	a.fmPending[reqID] = respCh
	a.mu.Unlock()

	defer func() {
		a.mu.Lock()
		delete(a.fmPending, reqID)
		a.mu.Unlock()
	}()

	if err := a.writeTextMessage(raw, 10*time.Second); err != nil {
		return "", nil, fmt.Errorf("发送 FM 请求失败: %w", err)
	}

	select {
	case resp := <-respCh:
		if resp.MsgType == msgTypeFMError {
			var errPayload fmErrorPayload
			if err := json.Unmarshal(resp.Payload, &errPayload); err == nil && errPayload.Message != "" {
				return "", nil, fmt.Errorf("%s", errPayload.Message)
			}
			return "", nil, fmt.Errorf("FM 操作失败")
		}
		return resp.MsgType, resp.Payload, nil
	case <-time.After(timeout):
		return "", nil, fmt.Errorf("FM 请求超时（%s）", timeout)
	case <-a.doneCh:
		return "", nil, fmt.Errorf("Agent 连接断开")
	}
}

// ── AgentHub FM 公开方法 ─────────────────────────────────────────────────────

// FMList 返回远端节点指定路径下的目录列表。
func (h *AgentHub) FMList(providerID uint, path string) (string, []FMEntry, error) {
	conn, ok := h.GetConn(providerID)
	if !ok || conn == nil {
		return "", nil, fmt.Errorf("agent 未连接")
	}

	_, payload, err := conn.sendFMRequest(msgTypeFMList, fmListPayload{Path: path}, fmDefaultTimeout)
	if err != nil {
		global.APP_LOG.Warn("FM List 失败", zap.Uint("providerID", providerID), zap.Error(err))
		return "", nil, err
	}

	var resp fmListRespPayload
	if err := json.Unmarshal(payload, &resp); err != nil {
		return "", nil, fmt.Errorf("解析 FM List 响应失败: %w", err)
	}
	return resp.Path, resp.Entries, nil
}

// FMDownload 下载远端节点指定路径的文件，返回文件字节内容。
func (h *AgentHub) FMDownload(providerID uint, path string) ([]byte, error) {
	conn, ok := h.GetConn(providerID)
	if !ok || conn == nil {
		return nil, fmt.Errorf("agent 未连接")
	}

	_, payload, err := conn.sendFMRequest(msgTypeFMDownload, fmDownloadPayload{Path: path}, fmDefaultTimeout)
	if err != nil {
		return nil, err
	}

	var resp fmDownloadRespPayload
	if err := json.Unmarshal(payload, &resp); err != nil {
		return nil, fmt.Errorf("解析 FM Download 响应失败: %w", err)
	}

	data, err := base64.StdEncoding.DecodeString(resp.Data)
	if err != nil {
		return nil, fmt.Errorf("解码文件内容失败: %w", err)
	}
	return data, nil
}

// FMUpload 上传文件到远端节点指定路径。
func (h *AgentHub) FMUpload(providerID uint, path string, data []byte) error {
	if len(data) > fmMaxUploadBytes {
		return fmt.Errorf("文件过大（最大 50 MB），请通过其他方式传输")
	}

	conn, ok := h.GetConn(providerID)
	if !ok || conn == nil {
		return fmt.Errorf("agent 未连接")
	}

	encoded := base64.StdEncoding.EncodeToString(data)
	_, _, err := conn.sendFMRequest(msgTypeFMUpload, fmUploadPayload{
		Path: path,
		Data: encoded,
		Size: int64(len(data)),
	}, fmUploadTimeout)
	return err
}

// FMDelete 删除远端节点指定路径的文件或空目录。
func (h *AgentHub) FMDelete(providerID uint, path string) error {
	conn, ok := h.GetConn(providerID)
	if !ok || conn == nil {
		return fmt.Errorf("agent 未连接")
	}

	_, _, err := conn.sendFMRequest(msgTypeFMDelete, fmDeletePayload{Path: path}, fmDefaultTimeout)
	return err
}

// FMMkdir 在远端节点创建目录（含父目录）。
func (h *AgentHub) FMMkdir(providerID uint, path string) error {
	conn, ok := h.GetConn(providerID)
	if !ok || conn == nil {
		return fmt.Errorf("agent 未连接")
	}

	_, _, err := conn.sendFMRequest(msgTypeFMMkdir, fmMkdirPayload{Path: path}, fmDefaultTimeout)
	return err
}

File: server/agent/src/fm.rs
// fm.rs — Agent 侧文件管理器实现。
// 处理来自控制端的 fm_list / fm_download / fm_upload / fm_delete / fm_mkdir 请求，
// 通过 WebSocket 控制通道返回文件操作结果。
//
// 协议与 Go 控制端 service/agent/fm.go 对称：
//   controller → agent: fm_list    { id, payload: {path} }
//   controller → agent: fm_download { id, payload: {path} }
//   controller → agent: fm_upload   { id, payload: {path, data(base64), size} }
//   controller → agent: fm_delete   { id, payload: {path} }
//   controller → agent: fm_mkdir    { id, payload: {path} }
//
//   agent → controller: fm_list_resp / fm_download_resp / fm_upload_resp /
//                        fm_delete_resp / fm_mkdir_resp  (成功)
//   agent → controller: fm_error   { id, payload: {message} }  (失败)

use base64::{Engine, engine::general_purpose::STANDARD as B64};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::path::Path;
use std::time::{Duration, UNIX_EPOCH};
use tokio::sync::mpsc;
use tokio_tungstenite::tungstenite::Message;
use tracing::{info, warn};

use crate::tunnel::WsFrame;

/// 最大下载文件大小（50 MB），保护 WebSocket 通道不被大文件阻塞。
const FM_MAX_DOWNLOAD_BYTES: u64 = 50 * 1024 * 1024;
const FM_RESPONSE_SEND_TIMEOUT_SECS: u64 = 10;

// ── 本地 payload 结构体（仅在此模块使用）────────────────────────────────────

#[derive(Deserialize)]
struct FMListPayload {
    path: String,
}

#[derive(Deserialize)]
struct FMDownloadPayload {
    path: String,
}

#[derive(Deserialize)]
struct FMUploadPayload {
    path: String,
    data: String,
    size: i64,
}

#[derive(Deserialize)]
struct FMDeletePayload {
    path: String,
}

#[derive(Deserialize)]
struct FMMkdirPayload {
    path: String,
}

#[derive(Serialize)]
struct FMEntry {
    name: String,
    #[serde(rename = "isDir")]
    is_dir: bool,
    size: i64,
    #[serde(rename = "modTime")]
    mod_time: i64,
    mode: String,
}

#[derive(Serialize)]
struct FMListRespPayload {
    path: String,
    entries: Vec<FMEntry>,
}

#[derive(Serialize)]
struct FMDownloadRespPayload {
    data: String,
    size: i64,
}

#[derive(Serialize)]
struct FMOkRespPayload {}

#[derive(Serialize)]
struct FMErrorPayload {
    message: String,
}

// ── 辅助发送函数 ─────────────────────────────────────────────────────────────

async fn send_fm_frame(ws_tx: &mpsc::Sender<Message>, frame: WsFrame) {
    let text = match serde_json::to_string(&frame) {
        Ok(text) => text,
        Err(e) => {
            warn!("FM serialize response frame failed: {e}");
            return;
        }
    };
    let msg = Message::Text(text.into());
    if ws_tx.try_send(msg.clone()).is_err() {
        let _ = tokio::time::timeout(
            Duration::from_secs(FM_RESPONSE_SEND_TIMEOUT_SECS),
            ws_tx.send(msg),
        )
        .await;
    }
}

async fn send_fm_ok<P: Serialize>(
    ws_tx: &mpsc::Sender<Message>,
    resp_type: &str,
    req_id: &str,
    payload: &P,
) {
    let payload_val = match serde_json::to_value(payload) {
        Ok(v) => v,
        Err(e) => {
            warn!("FM serialize ok payload failed: {e}");
            return;
        }
    };
    let frame = WsFrame {
        msg_type: resp_type.to_string(),
        id: Some(req_id.to_string()),
        payload: Some(payload_val),
    };
    send_fm_frame(ws_tx, frame).await;
}

async fn send_fm_error(ws_tx: &mpsc::Sender<Message>, req_id: &str, msg: &str) {
    warn!("FM error (id={}): {}", req_id, msg);
    let payload_val = match serde_json::to_value(FMErrorPayload {
        message: msg.to_string(),
    }) {
        Ok(v) => v,
        Err(_) => return,
    };
    let frame = WsFrame {
        msg_type: "fm_error".to_string(),
        id: Some(req_id.to_string()),
        payload: Some(payload_val),
    };
    send_fm_frame(ws_tx, frame).await;
}

// ── 公开处理函数（由 ws_client/handler.rs 调用）──────────────────────────────

/// 处理 fm_list 请求：列出目录内容
pub async fn handle_fm_list(
    id: Option<String>,
    payload: Option<Value>,
    ws_tx: mpsc::Sender<Message>,
) {
    let req_id = id.unwrap_or_default();
    let p = match payload.and_then(|v| serde_json::from_value::<FMListPayload>(v).ok()) {
        Some(p) => p,
        None => {
            send_fm_error(&ws_tx, &req_id, "invalid fm_list payload").await;
            return;
        }
    };
    info!(id = %req_id, path = %p.path, "fm_list");

    let dir_path = if p.path.is_empty() {
        "/".to_string()
    } else {
        p.path
    };

    // 尝试读目录；若失败，回退到 HOME
    let actual_path = match tokio::fs::metadata(&dir_path).await {
        Ok(m) if m.is_dir() => dir_path,
        _ => std::env::var("HOME").unwrap_or_else(|_| "/root".to_string()),
    };

    let mut read_dir = match tokio::fs::read_dir(&actual_path).await {
        Ok(rd) => rd,
        Err(e) => {
            send_fm_error(&ws_tx, &req_id, &format!("read dir failed: {e}")).await;
            return;
        }
    };

    let mut entries = Vec::new();
    while let Ok(Some(entry)) = read_dir.next_entry().await {
        let name = entry.file_name().to_string_lossy().to_string();
        if let Ok(meta) = entry.metadata().await {
            let mod_time = meta
                .modified()
                .ok()
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_millis() as i64)
                .unwrap_or(0);
            let mode = {
                #[cfg(unix)]
                {
                    use std::os::unix::fs::PermissionsExt;
                    format!("{:o}", meta.permissions().mode())
                }
                #[cfg(not(unix))]
                {
                    "0000".to_string()
                }
            };
            entries.push(FMEntry {
                name,
                is_dir: meta.is_dir(),
                size: meta.len() as i64,
                mod_time,
                mode,
            });
        }
    }

    send_fm_ok(
        &ws_tx,
        "fm_list_resp",
        &req_id,
        &FMListRespPayload {
            path: actual_path,
            entries,
        },
    )
    .await;
}

/// 处理 fm_download 请求：读取文件内容并以 base64 返回
pub async fn handle_fm_download(
    id: Option<String>,
    payload: Option<Value>,
    ws_tx: mpsc::Sender<Message>,
) {
    let req_id = id.unwrap_or_default();
    let p = match payload.and_then(|v| serde_json::from_value::<FMDownloadPayload>(v).ok()) {
        Some(p) => p,
        None => {
            send_fm_error(&ws_tx, &req_id, "invalid fm_download payload").await;
            return;
        }
    };
    info!(id = %req_id, path = %p.path, "fm_download");

    let meta = match tokio::fs::metadata(&p.path).await {
        Ok(m) => m,
        Err(e) => {
            send_fm_error(&ws_tx, &req_id, &format!("stat failed: {e}")).await;
            return;
        }
    };
    if meta.is_dir() {
        send_fm_error(&ws_tx, &req_id, "cannot download a directory").await;
        return;
    }
    if meta.len() > FM_MAX_DOWNLOAD_BYTES {
        send_fm_error(&ws_tx, &req_id, "file too large (max 50 MB)").await;
        return;
    }

    let data = match tokio::fs::read(&p.path).await {
        Ok(d) => d,
        Err(e) => {
            send_fm_error(&ws_tx, &req_id, &format!("read failed: {e}")).await;
            return;
        }
    };
    let size = data.len() as i64;
    send_fm_ok(
        &ws_tx,
        "fm_download_resp",
        &req_id,
        &FMDownloadRespPayload {
            data: B64.encode(&data),
            size,
        },
    )
    .await;
}

/// 处理 fm_upload 请求：解码 base64 写入指定路径
pub async fn handle_fm_upload(
    id: Option<String>,
    payload: Option<Value>,
    ws_tx: mpsc::Sender<Message>,
) {
    let req_id = id.unwrap_or_default();
    let p = match payload.and_then(|v| serde_json::from_value::<FMUploadPayload>(v).ok()) {
        Some(p) => p,
        None => {
            send_fm_error(&ws_tx, &req_id, "invalid fm_upload payload").await;
            return;
        }
    };
    info!(id = %req_id, path = %p.path, size = %p.size, "fm_upload");

    let data = match B64.decode(&p.data) {
        Ok(d) => d,
        Err(e) => {
            send_fm_error(&ws_tx, &req_id, &format!("base64 decode failed: {e}")).await;
            return;
        }
    };
    if data.len() as u64 > FM_MAX_DOWNLOAD_BYTES {
        send_fm_error(&ws_tx, &req_id, "file too large (max 50 MB)").await;
        return;
    }
    if p.size >= 0 && data.len() as i64 != p.size {
        send_fm_error(&ws_tx, &req_id, "upload size mismatch").await;
        return;
    }
    if let Some(parent) = Path::new(&p.path).parent() {
        if !parent.as_os_str().is_empty() {
            if let Err(e) = tokio::fs::create_dir_all(parent).await {
                send_fm_error(&ws_tx, &req_id, &format!("create parent dir failed: {e}")).await;
                return;
            }
        }
    }
    match tokio::fs::write(&p.path, &data).await {
        Ok(_) => send_fm_ok(&ws_tx, "fm_upload_resp", &req_id, &FMOkRespPayload {}).await,
        Err(e) => send_fm_error(&ws_tx, &req_id, &format!("write failed: {e}")).await,
    }
}

/// 处理 fm_delete 请求：删除文件或空目录
pub async fn handle_fm_delete(
    id: Option<String>,
    payload: Option<Value>,
    ws_tx: mpsc::Sender<Message>,
) {
    let req_id = id.unwrap_or_default();
    let p = match payload.and_then(|v| serde_json::from_value::<FMDeletePayload>(v).ok()) {
        Some(p) => p,
        None => {
            send_fm_error(&ws_tx, &req_id, "invalid fm_delete payload").await;
            return;
        }
    };
    info!(id = %req_id, path = %p.path, "fm_delete");

    let meta = match tokio::fs::metadata(&p.path).await {
        Ok(m) => m,
        Err(e) => {
            send_fm_error(&ws_tx, &req_id, &format!("stat failed: {e}")).await;
            return;
        }
    };
    let result = if meta.is_dir() {
        tokio::fs::remove_dir(&p.path).await
    } else {
        tokio::fs::remove_file(&p.path).await
    };
    match result {
        Ok(_) => send_fm_ok(&ws_tx, "fm_delete_resp", &req_id, &FMOkRespPayload {}).await,
        Err(e) => send_fm_error(&ws_tx, &req_id, &format!("delete failed: {e}")).await,
    }
}

/// 处理 fm_mkdir 请求：递归创建目录
pub async fn handle_fm_mkdir(
    id: Option<String>,
    payload: Option<Value>,
    ws_tx: mpsc::Sender<Message>,
) {
    let req_id = id.unwrap_or_default();
    let p = match payload.and_then(|v| serde_json::from_value::<FMMkdirPayload>(v).ok()) {
        Some(p) => p,
        None => {
            send_fm_error(&ws_tx, &req_id, "invalid fm_mkdir payload").await;
            return;
        }
    };
    info!(id = %req_id, path = %p.path, "fm_mkdir");

    match tokio::fs::create_dir_all(&p.path).await {
        Ok(_) => send_fm_ok(&ws_tx, "fm_mkdir_resp", &req_id, &FMOkRespPayload {}).await,
        Err(e) => send_fm_error(&ws_tx, &req_id, &format!("mkdir failed: {e}")).await,
    }
}
